### PR TITLE
Implemented basic derived init

### PR DESCRIPTION
**Pull request by hpux735**
This will allow the user to create a new Gaussian (though others could be added using the trait) from an iterator.  The distribution could then be used to create new samples that match the provided distribution.

### CODE REVIEW

File: src/distribution/gaussian.rs [hunk 1]
```diff
@@ -298,6 +300,35 @@ pub fn sample<S: Source>(source: &mut S) -> f64 {
...
 
+impl FromIter for Gaussian {
+    fn from_iter(iter: impl Iterator<Item=f64> + Clone) -> Self {
+        use core::f64::consts::PI;
+
+        let mut count = 0;
+        let mut mu = 0.0;
+        let mut sigma = 0.0;
+
+        // Compute the mean
```
Comment (IvanUkhov) [inline src/distribution/gaussian.rs:311]: Could you please remove the comments? It is self-explanatory.

File: src/distribution/gaussian.rs [hunk 2]
```diff
@@ -298,6 +300,35 @@ pub fn sample<S: Source>(source: &mut S) -> f64 {
...
+        for v in iter.clone() {
+            count += 1;
+            mu += v;
+        }
+
```
Comment (IvanUkhov) [inline src/distribution/gaussian.rs:316]: Please remove this one to have one block dedicated to the mean.

File: src/distribution/gaussian.rs [hunk 3]
```diff
@@ -298,6 +300,35 @@ pub fn sample<S: Source>(source: &mut S) -> f64 {
...
+        mu = mu / count as f64;
+
+        // Re-scan the iterator to compute the variance
+        for v in iter {
+            sigma += f64::powf(v - mu as f64, 2.0);
+        }
+
```
Comment (IvanUkhov) [inline src/distribution/gaussian.rs:323]: Please remove this one to have one block dedicated to the standard deviation.

File: src/distribution/gaussian.rs [hunk 4]
```diff
@@ -298,6 +300,35 @@ pub fn sample<S: Source>(source: &mut S) -> f64 {
...
+        // Re-scan the iterator to compute the variance
+        for v in iter {
+            sigma += f64::powf(v - mu as f64, 2.0);
+        }
+        
+        // Convert to the standard deviation
+        sigma = f64::sqrt(sigma / (count - 1) as f64);
+
+        // Return the new distribution
+        Gaussian {mu, sigma, norm: (2.0 * PI).sqrt() * sigma}
```
Comment (IvanUkhov) [inline src/distribution/gaussian.rs:328]: Please run `cargo fmt`.

Comment (IvanUkhov) [reply src/distribution/gaussian.rs:328]: Please call `Gaussian::new` instead to avoid the repetition.

File: src/distribution/gaussian.rs [hunk 5]
```diff
@@ -414,10 +445,26 @@ mod tests {
...
     use prelude::*;
 
+    use crate::distribution::FromIter;
+
     macro_rules! new(
         ($mu:expr, $sigma:expr) => (Gaussian::new($mu, $sigma));
     );
 
+    #[test]
+    fn iter() {
```
Comment (IvanUkhov) [inline src/distribution/gaussian.rs:455]: Please keep the function alphabetically sorted.

Comment (IvanUkhov) [reply src/distribution/gaussian.rs:455]: Please rename to `from_iter`.

File: src/distribution/gaussian.rs [hunk 6]
```diff
@@ -414,10 +445,26 @@ mod tests {
...
+        // Create a sample distribution to create an iterator from
```
Comment (IvanUkhov) [inline src/distribution/gaussian.rs:456]: Please remove the comment.

File: src/distribution/gaussian.rs [hunk 7]
```diff
@@ -414,10 +445,26 @@ mod tests {
...
+        let mut source = source::default(42);
+        let distribution = new!(1.0, 2.0);
+        let sampler = Independent(&distribution, &mut source);
+        let samples = sampler.take(10000).collect::<Vec<_>>();
+        let derived_distribution = Gaussian::from_iter(samples.into_iter());
+
+        assert::close(derived_distribution.mu, distribution.mu, 0.1);
```
Comment (IvanUkhov) [inline src/distribution/gaussian.rs:463]: The last one is epsilon. It is way to large. Can you please set it to, say, `1e-8`?

File: src/distribution/mod.rs [hunk 8]
```diff
@@ -99,6 +99,14 @@ pub trait Variance: Mean {
     }
 }
 
+/// A distribution capable of being initalized from an iterator
```
Comment (IvanUkhov) [inline src/distribution/mod.rs:102]: Please add a period at the end.

File: src/distribution/mod.rs [hunk 9]
```diff
@@ -99,6 +99,14 @@ pub trait Variance: Mean {
...
+/// 
+/// The trait is applicable when it's possible to infer the
```
Comment (IvanUkhov) [inline src/distribution/mod.rs:104]: Please change “it's” to “it is”. It is 80-character long, right?

File: src/distribution/mod.rs [hunk 10]
```diff
@@ -99,6 +99,14 @@ pub trait Variance: Mean {
...
+/// distribution parameters from a sample dataset.
+pub trait FromIter: Distribution {
+    fn from_iter(iter: impl Iterator<Item=f64> + Clone) -> Self;
```
Comment (IvanUkhov) [inline src/distribution/mod.rs:107]: This requires a docstring too.

**Review by IvanUkhov — COMMENTED** [review]
By the way, is there no standard train for constructing from an iterator? If `FromIter` is not something standard, one could maybe find a name matching the other ones. Like, `Variance` means it has a variance, and `Inverse` means it can be inverted. `Inference` could mean it can be inferred. Or `Inferable` to mimic `Continuous`.

Another thought is that this is a frequentist approach, MLE. What if, in the future, we want to do Bayesian? Maybe one could find a name to capture that. Maybe I overthink. One can always change later on.

**Comment by hpux735** [comment]
I have no opinion about the name.  I chose one just for the purposes of a placeholder.  As this is your package and project I'm happy to go with whatever you prefer.  Once we get that nailed down, I'll address the other comments. 

Thanks!

**Comment by IvanUkhov** [comment]
There is actually a standard trait for this:

https://doc.rust-lang.org/std/iter/trait.FromIterator.html

Let us just use it.

File: src/distribution/gaussian.rs (hunk 7 above)
Comment (hpux735) [reply src/distribution/gaussian.rs]: There's a reason it's so large.  If you want to have a small epsilon in this function, you'll need to make a very, very large set of random data to exactly match.  1e-8 could require a billion samples or more.  I tried generating with 100,000,000 samples (which makes the test take a few seconds) and only got this close:

`'1.0000279833901966 !~ 1.0'`

Comment (IvanUkhov) [reply src/distribution/gaussian.rs]: Ah, I see. Makes sense.

**Comment by IvanUkhov** [comment]
Looks very neat! Thank you!

**Comment by hpux735** [comment]
Awesome, thanks for the library!

**Pull request merged**